File: sorting/BubbleSort.java
package sorting;

import java.io.PrintWriter;
import java.io.StringWriter;

import fastIO.ArrayIO;
import fastIO.BasicIO;

/*
 * @Author: Niraj Thagunna
 * 
 * Bubble Sorting Algorithm - Implementation
 */
public class BubbleSort {
	
	// method to sort an array using Bubble Sorting Algorithm
	public static void bubbleSort(int[] arr, int n) {
		
		// loop for the number of passes
		for (int i = 0; i < n - 1; i++) {
			
			boolean flag = false;
			
			// loop for the iteration or comparing the pairs of adjacent elements
			for (int j = 0; j < n - 1 - i; j++) {
				
				if (arr[j] > arr[j + 1]) {
					int temp = arr[j];
					arr[j] = arr[j + 1];
					arr[j + 1] = temp;
					flag = true;
				}
			}
			
			// if no-swapping then break the loop
			if (flag == false)
				break;
		}
	}
	
	// main() or driver method
	public static void main(String[] args) {
		try {
			System.out.print("How many elements do you want to store? ");
			int MAX = BasicIO.nextInt();
			
			// Creating an array
			int[] arr = ArrayIO.readArray(MAX);
			
			System.out.println("\nOriginal Array : ");
			ArrayIO.printArray(arr);
			
			// calling the bubble sort algorithm
			bubbleSort(arr, MAX);
			System.out.println("\nArray after sorting - Bubble Sort : ");
			ArrayIO.printArray(arr);
		}
		catch (Exception ex) {
			StringWriter sw = new StringWriter();
			PrintWriter pw = new PrintWriter(sw);
			ex.printStackTrace(pw);
			String error = sw.toString();
			System.out.println("Error : \n" + error);
		}
	}
}

File: sorting/bubblesort.cpp
#include <iostream>
using namespace std;

// Bubble Sort is the simplest sorting algorithm that works by repeatedly swapping 
// the adjacent elements if they are in wrong order.

void bubbleSort(int a[], int n)
{
    int flag;
    for (int i = 0; i < n - 1; i++)
    {
        flag = 0;
        for (int j = 0; j < n - 1 - i; j++)
        {
            if (a[j] > a[j + 1])
            {
                swap(a[j], a[j + 1]);
                flag = 1;
            }
        }
        if (flag == 0)
        {
            break; // that means the array was already sorted
        }
    }
    // Time Complexity : O(n*n) for average and worst case while O(n) for the best case
    // Best case means the array was already sorted
}

int main()
{

    int n;
    cout << "\n Please enter the no. of elements in the array: ";
    cin >> n;
    int arr[n];
    cout << "\n Please enter the elements of the array: ";
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }
    bubbleSort(arr, n);
    cout << "\n After Sorting: \n";
    for (int i = 0; i < n; i++)
    {
        cout << " " << arr[i] << " ";
    }
    return 0;
}